checker/jdk/nullness/src/java/util/concurrent/CopyOnWriteArrayList.java

```diff
@@ -0,0 +1,1698 @@
+                setArray(newElements);
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Appends the element, if not present.
+     *
+     * @param e element to be added to this list, if absent
+     * @return {@code true} if the element was added
+     */
+    public boolean addIfAbsent(E e) {
+        Object[] snapshot = getArray();
+        return indexOf(e, snapshot, 0, snapshot.length) >= 0 ? false :
+            addIfAbsent(e, snapshot);
+    }
+
+    /**
+     * A version of addIfAbsent using the strong hint that given
+     * recent snapshot does not contain e.
+     */
+    private boolean addIfAbsent(E e, Object[] snapshot) {
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            Object[] current = getArray();
+            int len = current.length;
+            if (snapshot != current) {
+                // Optimize for lost race to another addXXX operation
+                int common = Math.min(snapshot.length, len);
+                for (int i = 0; i < common; i++)
+                    if (current[i] != snapshot[i] && eq(e, current[i]))
+                        return false;
+                if (indexOf(e, current, common, len) >= 0)
+                        return false;
+            }
+            Object[] newElements = Arrays.copyOf(current, len + 1);
+            newElements[len] = e;
+            setArray(newElements);
+            return true;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Returns {@code true} if this list contains all of the elements of the
+     * specified collection.
+     *
+     * @param c collection to be checked for containment in this list
+     * @return {@code true} if this list contains all of the elements of the
+     *         specified collection
+     * @throws NullPointerException if the specified collection is null
+     * @see #contains(Object)
+     */
+    public boolean containsAll(Collection<? extends @NonNull Object> c) {
+        Object[] elements = getArray();
+        int len = elements.length;
+        for (Object e : c) {
+            if (indexOf(e, elements, 0, len) < 0)
+                return false;
+        }
+        return true;
+    }
+
+    /**
+     * Removes from this list all of its elements that are contained in
+     * the specified collection. This is a particularly expensive operation
+     * in this class because of the need for an internal temporary array.
+     *
+     * @param c collection containing elements to be removed from this list
+     * @return {@code true} if this list changed as a result of the call
+     * @throws ClassCastException if the class of an element of this list
+     *         is incompatible with the specified collection
+     *         (<a href="../Collection.html#optional-restrictions">optional</a>)
+     * @throws NullPointerException if this list contains a null element and the
```

Comment:
Given how detailed the spec is about `NullPointerException` for `removeAll` and `retainAll` (and arguably even `containsAll`), I could see accepting collections with null elements. But it's a judgment call.

(_If_ you buy that, it would apply to <code>CopyOnWriteArray**Set**</code>, as well.)

Also, this prompted me to file https://github.com/typetools/checker-framework/issues/3197, which might be helpful as part of mounting an argument for `Collection.contains` and friends to always accept `null`.

Reply:
You are right (in #3197) that the current annotations are unsound.  Let's put that aside for the moment, because it requires a fix for the "annotations on receiver type parameters are ignored" problem.

I think this needs to stand as is in the annotated JDK, conservatively requiring non-null element types (the second condition in #3197).  Our goal is that if the client code satisfies the annotations, then there is no possibility of a NullPointerException.
The fix in #3197 would be more precise.

Does that make sense?  Or am I missing something?

Reply:
Having thought about this more in #3197, I think the more conservative option is actually to _require_ a nullable element type for the parameter. That ensures that `copyOnWriteArraySet.removeAll(c)` can safely call `c.contains(null)`.

But I'm not sure if you want to go back and tweak a bunch of `removeAll` methods at this point, especially if you'd just be re-tweaking them after annotations on receiver type parameters start to work.

Reply:
I would like to leave this until after annotations on receiver type arguments are no longer ignored.  Thanks for raising it, though.